### PR TITLE
Bug 1241305 - test contactslist favlist

### PR DESCRIPTION
**Pull request by kglazko**
Adding a JS Test to the Contacts details_test.js to check whether a favorited contact will show in the regular contacts list and the favorites list.

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bafc470ecb419b58311df4f065b35a4225238bf5

File: apps/communications/contacts/test/marionette/details_test.js [hunk 1]
```diff
@@ -110,4 +110,28 @@ marionette('Contacts > Details', function() {
     var menuOptions = sysMenu.findElements('button');
     assert.ok(menuOptions.length >= 2);
   });
+
+  test('Fav Contact in Reg/Fav List', function() {
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/details_test.js]: I'd leave a more self-explanatory sentence. Something like: 'Favorite contact should appear in favorite and regular lists'

As we already have another test for favorites, we can use [mocha's TDD APIs](http://mochajs.org/#tdd) more efficiently. For instance:
```js
suite('Favorite contact', function() {
  test('should be marked as favorite in details view', function() {
    // test
  });

  test('should appear in favorite and regular lists', function() {
    // test
  });
})
```

File: apps/communications/contacts/test/marionette/details_test.js [hunk 2]
```diff
@@ -110,4 +110,28 @@ marionette('Contacts > Details', function() {
...
+    var nameNode = client.helper.waitForElement(selectors.detailsHeader);
+    client.waitFor(function() {
+      return nameNode.getAttribute('class').indexOf('favorite') != -1;
+    });
+    //Wait for return to Contacts List
+    subject.backToList();
+
+    //Verify Contact is in Regular Contacts List
+    var cName = client.findElement(selectors.listContactFirstText);
+    assert.ok(cName.displayed(), 'Element should exist.');
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/details_test.js]: As a rule of thumb: use custom messages [only when it adds more information to the error](https://wiki.mozilla.org/B2G/QA/Automation/Style_Guide/Best_Practices#Assertion_Messages).

Unlike the example given in the link, here we don't hide any data here. So it's not a real problem. Nevertheless, it's a good habit to have, and it helps in having a consistent code base.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7ee6ff3726187b478c1a08dfabed0e624b30e9a6

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5c2ef22e4513364bd6e49524735117d02a030f16

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=13b31bfb4db00293b8fbf02f344a975a96afd711

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=40cc8dc632e60c5b090cfa5c251bbac7de331b8a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7d9811d1fd23e61900488a7c06db409e33dac332

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2289a502bd15fbc3e11f88c0366d358f72b8b30d

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=76bba9c087ad554e22d3e85d8268defa01df7a4f

File: apps/communications/contacts/test/marionette/details_test.js [hunk 3]
```diff
@@ -3,6 +3,7 @@
 var Contacts = require('./lib/contacts');
 var ContactsData = require('./lib/contacts_data');
 var assert = require('assert');
+var ContactsListView = require('./lib/views/contact-list/view.js');
```
Comment (mwargers) [inline apps/communications/contacts/test/marionette/details_test.js]: The .js part is needed here? Normally `require` seems to be used without .js.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7a7a3bf0edbe9a70ddd7ef7c63257dac4c4c350e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=768ec714fef6ebc5626759952bc9865f0c0e28cf

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=59edb259d7eb5ba4afdbe9f1b0bf4c48d4eb4d34

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=46a04b498b75e929b8fa62b17db7a489161c1647

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9b781bb04b696e69d6ad81f6d81d622c84d25edf

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ee0f32c18c86969937cb69d276f1306ba9255e5f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=15eec3428cb21d163ecc9fc15449d47900991fca

**Comment by yglazko** [comment]
One thing to go- need to write my own 'Subject' materials within the view! Once I get rid of the subject dependency, and format, it should be good to go.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9b0ec96956bd099fc14f7e24b22e4490feafb68c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dcd512a0810ef56b4007552438eb7eee7902ecf7

**Comment by yglazko** [comment]
Finished with removing subject dependencies

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=94bcf945e5249bd47a9a1e668fce1d4bd8ad202a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=14213f7ef30e97d1671e9b439d611e787feea738

File: apps/communications/contacts/test/marionette/details_test.js [hunk 4]
```diff
@@ -109,5 +90,27 @@ marionette('Contacts > Details', function() {
...
     var menuOptions = sysMenu.findElements('button');
     assert.ok(menuOptions.length >= 2);
+  }); 
+
+  suite('Favorite contact', function() {
+    setup(function(){
+      contactsData.createMozContact(testContact);
+    });
+    test('Mark contact as favorite', function() {
+      var contactsListView = new ContactsListView(client, subject);
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/details_test.js]: Nit: `subject` is not needed anymore.

File: apps/communications/contacts/test/marionette/details_test.js [hunk 5]
```diff
@@ -109,5 +90,27 @@ marionette('Contacts > Details', function() {
...
+      var contactDetailsView = contactsListView.goToContact();
+      contactDetailsView.makeFavorited();
+      assertContactData(testContact);
+    });
+
+    test('Favorite contact should appear in' +
+      ' favorite and regular lists', function() {
+      var contactsListView = new ContactsListView(client, subject);
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/details_test.js]: Nit: `subject` is not needed anymore.

File: apps/communications/contacts/test/marionette/lib/views/contact-details/view.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      		var pointerEvents = element.cssProperty('pointer-events');
+			return opacity == 1 && pointerEvents == 'auto';
+    	};
+    	this.client.waitFor(test);
+  	},
+
+	backtoContactsList: function() {
+		this.backToList();
+		var ContactsListView = require('../contact-list/view');
+		var contactsListView = new ContactsListView(this.client, this.subject);
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/lib/views/contact-details/view.js]: Nit: `this.subject` is not needed anymore.

File: apps/communications/contacts/test/marionette/lib/views/contact-details/view.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+function ContactsDetailsView(client, subject) {
+	this.client = client;
+	this.subject = subject;
+	this.accessors = new ContactsDetailsAccessor(client);
+}
+
+ContactsDetailsView.prototype = {
+	makeFavorited: function() {
+		this.accessors.favoritesButton.tap();
+		return this.waitForFavorited;
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/lib/views/contact-details/view.js]: :warning: No need to return the function itself. Calling `this.waitForFavorited()` will do the job.

[More details here](https://github.com/yglazko/gaia/commit/4479b74afae0fcaed81d32d7cf05bc0468566b4f#commitcomment-15614634).

File: apps/communications/contacts/test/marionette/lib/views/contact-details/view.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	},
+
+	waitForFavorited: function() {
+		function waitFav(temp) {
+        	return temp.detailsHeader.getAttribute('class'
+        	).indexOf('favorite') != -1;
+      	}
+      	this.client.waitFor(waitFav(this.accessors));
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/lib/views/contact-details/view.js]: :warning: This line should have a 2-space-indentation as it doesn't belong to `waitFav`.

File: apps/communications/contacts/test/marionette/lib/views/contact-details/view.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+ContactsDetailsView.prototype = {
+	makeFavorited: function() {
+		this.accessors.favoritesButton.tap();
+		return this.waitForFavorited;
+	},
+
+	waitForFavorited: function() {
+		function waitFav(temp) {
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/lib/views/contact-details/view.js]: We can make this signature more self-documented. How about:
``` js
var self = this;
function waitForHeaderToContainFavoriteClass() {
  return self.detailsHeader.getAttribute('class').indexOf('favorite') != -1;
}
```

`self = this` is a common way to conserve `this` in an inner scope, in JS.

Also, if you think `waitForHeaderToContainFavoriteClass` is too redundant, you can inline this way:
```js
var self = this;
this.client.waitFor(function() {
  return self.detailsHeader.getAttribute('class').indexOf('favorite') != -1;
}
```

File: apps/communications/contacts/test/marionette/lib/views/contact-details/view.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        	return temp.detailsHeader.getAttribute('class'
+        	).indexOf('favorite') != -1;
+      	}
+      	this.client.waitFor(waitFav(this.accessors));
+	},
+
+	backToList: function() {
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/lib/views/contact-details/view.js]: Okay to duplicate this function from `contacts.js` now. Nonetheless, I'd leave a TODO comment say that this clean needs to be done by creating a ContactsViewBase class. Then ContactsDetailsView would inherit from it.

File: apps/communications/contacts/test/marionette/lib/views/contact-list/view.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+	get firstFavorite() {
+		return this.accessors.firstFavorite;
+	},
+
+	waitSlideLeftDetails: function() {
+		var element = this.client.findElement(
+		this.accessors.detailsSelector),
+        location;
+    	var test = function() {
```
Comment (JohanLorenzo) [inline apps/communications/contacts/test/marionette/lib/views/contact-list/view.js]: [Like there](https://github.com/mozilla-b2g/gaia/pull/33928/files#r50830027), we can be more explicit about the content of this `test`. How about: `waitUntilPanelReachedLeftBorder()` ?

If you think it's too redundant, we can inline it.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d4456302e7f71fe2fd91127e3d578af1e52053f4

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=16b12370a627772febf3adef39cf7ffe68818705

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=be3270ba04c6e7249a64a70c5ab11f553437ba64

**Pull request merged**